File: 49. Group Anagrams/49. Group Anagrams.py
from collections import defaultdict

class Solution(object):
    def groupAnagrams(self, strs):
        """
        :type strs: List[str]
        :rtype: List[List[str]]
        """
        anagram_groups = defaultdict(list)
        
        for s in strs:
            # Create a tuple of character counts to use as a key
            key = tuple(sorted(self.char_count(s)))
            anagram_groups[key].append(s)
        
        return list(anagram_groups.values())
    
    def char_count(self, s):
        """
        Helper function to count characters in a string.
        Returns a list of 26 integers, each representing the count of a lowercase letter.
        """
        count = [0] * 26
        for char in s:
            count[ord(char) - ord('a')] += 1
        return count

File: 49. Group Anagrams/49. Group Anagrams.rs
use std::collections::HashMap;

impl Solution {
    pub fn group_anagrams(strs: Vec<String>) -> Vec<Vec<String>> {
        let mut anagram_groups: HashMap<Vec<char>, Vec<String>> = HashMap::new();
        
        for s in strs {
            let mut sorted_chars: Vec<char> = s.chars().collect();
            sorted_chars.sort_unstable();
            anagram_groups.entry(sorted_chars).or_insert(Vec::new()).push(s);
        }
        
        anagram_groups.into_values().collect()
    }
}